kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandEncryptionManager.java

```diff
@@ -0,0 +1,263 @@
+                    return CompletableFuture.completedFuture(encrypt);
+                }
+                catch (ExhaustedDekException e) {
+                    rotateKeyContext(encryptionScheme, dek);
+                    // fall through to recursive call below...
+                }
+                catch (Exception e) {
+                    return CompletableFuture.failedFuture(e);
+                }
+            }
+            // recurse, incrementing the attempt number
+            return attemptEncrypt(topicName, partition, encryptionScheme, records, attempt + 1, batchRecordCounts, bufferAllocator);
+        });
+    }
+
+    @NonNull
+    private BatchAwareMemoryRecordsBuilder encryptBatches(@NonNull EncryptionScheme<K> encryptionScheme,
+                                                          @NonNull MemoryRecords memoryRecords,
+                                                          @NonNull Dek.Encryptor encryptor,
+                                                          @NonNull IntFunction<ByteBufferOutputStream> bufferAllocator) {
+        BatchAwareMemoryRecordsBuilder builder = new BatchAwareMemoryRecordsBuilder(allocateBufferForEncrypt(memoryRecords, bufferAllocator));
+        for (MutableRecordBatch batch : memoryRecords.batches()) {
+            maybeEncryptBatch(encryptionScheme, encryptor, batch, builder);
+        }
+        return builder;
+    }
+
+    private void maybeEncryptBatch(@NonNull EncryptionScheme<K> encryptionScheme,
+                                   @NonNull Dek.Encryptor encryptor,
+                                   @NonNull MutableRecordBatch batch,
+                                   @NonNull BatchAwareMemoryRecordsBuilder builder) {
+        if (batch.isControlBatch()) {
+            builder.writeBatch(batch);
+        }
+        else {
+            List<Record> records = StreamSupport.stream(batch.spliterator(), false).toList();
+            if (records.isEmpty()) {
+                builder.writeBatch(batch);
+            }
+            else {
+                // TODO per-thread caching
+                ByteBuffer recordBuffer = ByteBuffer.allocate(recordBufferInitialBytes);
+                do {
+                    try {
+                        RecordBatchUtils.toMemoryRecords(batch,
+                                new RecordEncryptor<>(encryptionVersion, encryptionScheme, encryptor, dekManager.edekSerde(), recordBuffer),
+                                builder);
+                        break;
+                    }
+                    catch (BufferTooSmallException e) {
+                        int newCapacity = 2 * recordBuffer.capacity();
+                        if (newCapacity > recordBufferMaxBytes) {
+                            throw new EncryptionException("Record buffer cannot grow greater than " + recordBufferMaxBytes + " bytes");
+                        }
+                        recordBuffer = ByteBuffer.allocate(newCapacity);
```

Comment:
Yes, there will be an effect on tail latency. Who knows whether it's really measurable (as in distinguishable from the noise).

We almost certainly do want metrics, for this and things like cache usage (maybe Caffeine provides some itself). But we need to figure out what metrics are going to be useful to people operating this thing (rather than metrics which are just kinda interesting). I think this is easy to punt to a later PR (but we should think carefully whether we want some metrics in 0.5.0 though).